### PR TITLE
✨(frontend) order details: add contract

### PR DESCRIPTION
**Pull request by rlecellier**
## Purpose

When a product with contract_definition is purchased,
The learner have to sign the order.contract before
enrolling to training's courses.
If the contract have been signed, the learner can
download it.

There a new section in order details page that
handle these actions.

### Listing
image

### Signature required
image

## Contract signed
image

### CODE REVIEW

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx [hunk 1]
```diff
@@ -79,34 +152,43 @@ export const DashboardItemOrder = ({
...
+                  className="dashboard-item__button"
+                  color="transparent-darkest"
+                  href={getRoutePath(LearnerDashboardPaths.ORDER, { orderId: order.id })}
+                  data-testid="dashboard-item-order__button"
+                >
+                  {intl.formatMessage(messages.accessCourse)}
+                </RouterButton>
+              )}
+            </div>
+            {!writable && !order.contract?.signed_on && (
```
Comment (NathanVss) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx:186]: This condition could cause a bug. All order without a contract are showing the banner "contract to sign"

Capture d’écran 2023-10-24 à 17 01 10

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx [hunk 2]
```diff
@@ -216,7 +231,13 @@ const DashboardItemCourseEnrollingRun = ({
             <Icon name={IconTypeEnum.CHECK} size="small" />
           </div>
         ) : (
-          <Button disabled={!isOpenedForEnrollment} color="secondary" onClick={enroll}>
+          <Button
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:234]: My bad, yes they are

[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:234]: If the order is readonly AND needs a signature the `Enroll` button are not disabled in read only mode.

Capture d’écran 2023-10-25 à 11 49 37

[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:234]: So I did the following update: 
Capture d’écran 2023-10-25 à 11 57 45

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx [hunk 3]
```diff
@@ -216,7 +231,13 @@ const DashboardItemCourseEnrollingRun = ({
...
+            className="dashboard-item__button"
+            disabled={!isOpenedForEnrollment || haveToSignContract}
+            color="secondary"
+            onClick={enroll}
+            title={haveToSignContract ? intl.formatMessage(messages.contractUnsigned) : ''}
```
[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:245]: I added it in the last fixup

[the remark this answers is not in the document]
Comment (NathanVss) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:245]: Capture d’écran 2023-10-25 à 13 15 14

File: src/frontend/js/utils/test/factories/joanie.ts [hunk 4]
```diff
@@ -71,12 +72,23 @@ export const TargetCourseFactory = factory((): TargetCourse => {
...
+  };
+});
+
 export const ContractFactory = factory((): Contract => {
   return {
     id: faker.string.uuid(),
     learner_name: faker.person.fullName(),
     product_title: faker.lorem.words(10),
-    sign_date: faker.date.past().toISOString(),
+    signed_on: faker.date.past().toISOString(),
```
Comment (jbpenrath) [inline src/frontend/js/utils/test/factories/joanie.ts:91]: A contract object does not have this shape.

It should have the following : 
```ts
{
                "id": UUID,
                "created_on": Date,
                "signed_on": Date | undefined
                "definition": ContractDefinition,
                "order": {
                    "id": UUID,
                    "course": {
                        "code": String
                        "cover": JoanieFile,
                        "id": UUID,
                        "title": String,
                    },
                    "organization": {
                        "id": UUID,
                        "code": String,
                        "logo": JoanieFile,
                        "title": String,
                    },
                    "owner": String,
                    "product": String // The product title,
                },
            },
```

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/OrderStateMessage/index.tsx [hunk 5]
```diff
@@ -64,12 +64,6 @@ const OrderStateMessage = ({ order }: OrderStateMessageProps) => {
     [OrderState.CANCELED]: messages.statusCanceled,
   };
 
-  useEffect(() => {
-    if (!(order.state in orderStatusMessages)) {
-      handle(new Error(`Unknow order state ${order.state}`));
-    }
-  }, [order.state]);
-
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/OrderStateMessage/index.tsx:72]: You can revert this change. The PR #2124 fixes this bug.

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx [hunk 6]
```diff
@@ -92,6 +93,7 @@ export const DashboardItemCourseEnrolling = ({
   order,
   icon = false,
   notEnrolledUrl = '#',
+  hideEnrollButtons,
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:96]: `disableEnrollButton` is a better name WDYT ?

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx [hunk 7]
```diff
@@ -183,7 +183,7 @@ export const DashboardItemOrder = ({
                 </RouterButton>
               )}
             </div>
-            {!writable && !order.contract?.signed_on && (
+            {!writable && order.contract && !order.contract.signed_on && (
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx]: What do you think to store this condition statement within a variable with an explicit name like `hasContractToSign` ?

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx (hunk 6 above)
Comment (NathanVss) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:96]: I think it could be confusing because disabled could mean "display the button but turn them disabled", but in this scenario we want to completely hide them, nope?

Comment (jbpenrath) [reply src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:96]: My bad! I was confused by `NotEnrolled` and `DashboardItemCourseEnrollingRuns` components.
https://github.com/openfun/richie/pull/2119#discussion_r1371564619

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx [hunk 8]
```diff
@@ -69,13 +69,19 @@ const messages = {
     description: 'Text displayed when course runs list is loading',
     id: 'components.DashboardItemCourseEnrollingRun.courseRunsLoading',
   },
+  contractUnsigned: {
+    id: 'components.DashboardItemCourseEnrollingRun.contractUnsigned',
+    description: 'Message displayed as disabled button title when a contract needs to be signed.',
+    defaultMessage: 'You have to sign the training contrat before enrolling to your course.',
+  },
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/DashboardItemCourseEnrolling.tsx:76]:
```suggestion
  contractUnsigned: {
    id: 'components.DashboardItemCourseEnrolling.contractUnsigned',
    description: 'Message displayed as disabled button title when a contract needs to be signed.',
    defaultMessage: 'You have to sign the training contract before enrolling to your course.',
  },
```

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx [hunk 9]
```diff
@@ -26,6 +27,27 @@ const messages = {
     description: 'Accessible label displayed while certificate is being fetched on the dashboard.',
     defaultMessage: 'Loading certificate...',
   },
+  contractUnsigned: {
+    id: 'components.DashboardItemOrderContract.contractUnsigned',
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx:31]: To keep id consistent respect the naming convention `components.<COMPONENT_NAME>.messageKey`

```suggestion
    id: 'components.DashboardItemOrder.contractUnsigned',
```

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx [hunk 10]
```diff
@@ -26,6 +27,27 @@ const messages = {
...
+    description: "Message displayed when the order's contract needs to be signed.",
+    defaultMessage: 'You have to sign this contract to access your training.',
+  },
+  contractSigned: {
+    id: 'components.DashboardItemOrderContract.contractSigned',
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx:36]: Ditto

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx [hunk 11]
```diff
@@ -26,6 +27,27 @@ const messages = {
...
+    description:
+      "Message displayed when the order's contract has been signed and can be downloaded.",
+    defaultMessage: "You've accepted the training contract.",
+  },
+  contractSignActionLabel: {
+    id: 'components.DashboardItemOrderContract.contractSignActionLabel',
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx:42]: Ditto

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx [hunk 12]
```diff
@@ -26,6 +27,27 @@ const messages = {
...
+    description: 'Label of "sign contract" action.',
+    defaultMessage: 'Sign',
+  },
+  contractDownloadActionLabel: {
+    id: 'components.DashboardItemOrderContract.contractDownloadActionLabel',
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx:47]: Ditto

File: src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx [hunk 13]
```diff
@@ -40,6 +62,58 @@ interface DashboardItemOrderCertificateProps {
...
+  mode?: 'default' | 'compact';
+}
+
+const DashboardItemOrderContractFooter = ({ contract }: DashboardItemOrderContractFooterProps) => {
+  const signContract = () => {
+    alert('TODO: sign contract action');
+  };
+  const downloadContract = () => {
+    alert('TODO: download contract action');
+  };
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardItem/Order/DashboardItemOrder.tsx:81]: I would make a noop but add a TODO comment instead

File: src/frontend/js/widgets/Dashboard/components/DashboardOrderLoader/index.tsx [hunk 14]
```diff
@@ -13,6 +13,11 @@ const messages = defineMessages({
     description: 'Message displayed while loading an order',
     id: 'components.DashboardOrderLoader.loading',
   },
+  signatureNeeded: {
+    defaultMessage: 'You need to sign your contract before enrolling to your courses',
```
Comment (jbpenrath) [inline src/frontend/js/widgets/Dashboard/components/DashboardOrderLoader/index.tsx:17]:
```
    defaultMessage: 'You need to sign your contract before enrolling to course runs',
```

**Review by jbpenrath — APPROVED** [review]
LGTM ✅
